### DISCUSSION TITLE
why does term.ui need a `frame_rate`?

### DISCUSSION THREAD

**Discussion by emdete** [vlang/v#16196] [General]
i wonder why an library for terminal display uses a frame rate, normally characters should be changed on screen if the program decides so and not 30 times a second. isnt that a waste of resources if the characters dont change (which will be the case most of the time for usual TUIs). shouldnt such a lib use a select call to detect what to do next?

**Comment by JalonSolov** [comment]
It's there because you _can_ do animations in the terminal, and you want to be able to ensure they don't run too fast.  See `examples/term.ui/pong.v` and `examples/term.ui/vyper.v`.

**Comment by emdete** [comment]
but it wastes resources, right? without need the screen is updated 30 times a second without any change for usual terminal programs. in terms of power consumption (for example on a mobile device) this isnt desired.

**Reply by spytheman** [reply]
You can implement caching in your .frame() callback, so that it returns early, when there are no changes.

**Reply by emdete** [reply]
how would that help? i would like the process to 'hang' in an io wait until something happens that requires action. no cpu used until then. good old unix style ;)

**Reply by bogen85** [reply]
> how would that help? i would like the process to 'hang' in an io wait until something happens that requires action. no cpu used until then. good old unix style ;)

It might be possible to just wait in the .frame() callback until there is something ready to be done.

But it seems like you are really looking for is some wait events or wait inputs, or something similar.

I've not studied `term.ui` yet, so I don't know off hand.
